pkg/controller/master/nodedrain/nodedrain_controller.go

```diff
@@ -430,18 +430,38 @@ func (ndc *ControllerHandler) CheckVMISchedulingRequirements(originalNode *corev
 			// identify if nodeAffinity can be met by other nodes and node is ready
 			for _, v := range nodeList {
 				if nodeAffinitySelector.Match(v) && v.Name != originalNode.Name && isNodeReady(v) {
-					possibleNodes = append(possibleNodes, v.Name)
+					possibleNodes = append(possibleNodes, v)
 				}
 			}
+
+			validNodes = filterNodesForNodeSelector(possibleNodes, vmi)
 			// no valid node found that could meet the requirements
-			if len(possibleNodes) == 0 {
+			if len(validNodes) == 0 {
 				impactedVMS = append(impactedVMS, namespacedVMName(vmi))
 			}
 		}
 	}
 	return impactedVMS, nil
 }
 
+// filterNodesForNodeSelector will filter nodes for vmi node selector requirement match
+func filterNodesForNodeSelector(possibleNodes []*corev1.Node, vmi *kubevirtv1.VirtualMachineInstance) []*corev1.Node {
+	var validNodes []*corev1.Node
+	// VM's may also have node selector, which is used when defining specific hostnames
+	if vmi.Spec.NodeSelector != nil && len(vmi.Spec.NodeSelector) > 0 {
+		vmiNodeSelector := labels.SelectorFromSet(vmi.Spec.NodeSelector)
+		for _, v := range possibleNodes {
+			nodeLabels := labels.Set(v.GetLabels())
+			if vmiNodeSelector.Matches(nodeLabels) {
+				validNodes = append(validNodes, v)
+			}
+		}
```

Comment:
How about the following?
```suggestion
		}
		return validNodes
	}
	return possibleNodes
```

The `validNodes` will be filled and we can return if we get nodeSelector. Others return the `possibleNodes` 
WDYT?